### PR TITLE
added histogram to facet_grid

### PR DESCRIPTION
**Pull request by Kully**

### CODE REVIEW

**Comment by Kully** [comment]
screen shot 2017-06-20 at 12 38 26 pm

**Comment by Kully** [comment]
@chriddyp @cldougl 
need a review when tests pass

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/).
 
+## [Unreleased]
+### Added
+- `.create_facet_grid` now supports histogram traces.
```
Comment (cldougl) [inline CHANGELOG.md]: maybe `figure_factory.create_facet_grid` just to be super clear

**Comment by cldougl** [comment]
just tried putting hist in one of the docstring examples and it's rendered a bit weird and needs to be autoscaled:

```
import plotly.plotly as py
import plotly.figure_factory as ff  

import pandas as pd
mpg = pd.read_table('https://raw.githubusercontent.com/plotly/datasets/master/mpg_2017.txt')

fig = ff.create_facet_grid(mpg, x='displ', y='cty', facet_col='cyl', trace_type='histogram')
py.iplot(fig, filename='facet_grid_mpg_one_way_facet')
```

hist

**Comment by Kully** [comment]
Note: going to add lines to histogram for aesthetics

**Comment by Kully** [comment]
@cldougl tests finally passed. good to merge?

**Comment by cldougl** [comment]
I'll take another look this afternoon @Kully 
can you fix the changelog conflicts?

File: plotly/figure_factory/_facet_grid.py [hunk 2]
```diff
@@ -25,6 +25,7 @@
 THRES_FOR_FLIPPED_FACET_TITLES = 10
 GRID_WIDTH = 1
 
+VALID_TRACE_TYPES = ['scatter', 'scattergl', 'histogram']
```
Comment (chriddyp) [inline plotly/figure_factory/_facet_grid.py]: technically other types too though, right? I think that `box`, `histogram2d`, and `bar` might work now

File: plotly/figure_factory/_facet_grid.py [hunk 3]
```diff
@@ -189,18 +190,22 @@ def _facet_grid_color_categorical(df, x, y, facet_row, facet_col, color_name,
     if not facet_row and not facet_col:
         color_groups = list(df.groupby(color_name))
         for group in color_groups:
-            trace = graph_objs.Scatter(
+            trace = dict(
                 x=group[1][x],
                 y=group[1][y],
```
Comment (chriddyp) [inline plotly/figure_factory/_facet_grid.py]: I'm having a hard time understanding what happens when the `histogram` mode sets both `x` and `y` here. Shouldn't only one of them be set?

Comment (chriddyp) [inline plotly/figure_factory/_facet_grid.py]: Or is just the other variable ignored?

File: plotly/figure_factory/_facet_grid.py (hunk 2 above)
Comment (Kully) [inline plotly/figure_factory/_facet_grid.py]: I remember not liking the look of box or bar - it was very squashed up, but perhaps it was the dataset

File: plotly/figure_factory/_facet_grid.py (hunk 3 above)
Comment (chriddyp) [inline plotly/figure_factory/_facet_grid.py]: Oh, or the user just doesn't supply the `y` variable when setting it? Like 
```
FF.facet_grid(df, x='categorial-column', facet_col='another-column', trace_type='histogram')
```

Comment (Kully) [inline plotly/figure_factory/_facet_grid.py]: You found an oversight in my design.

Originally this was meant for just scatter, so I was forcing `x` and `y`. I'm going to try not forcing `x` and `y` and putting `df[x]` and `df[y]` in the trace dictionary depending on what the user provides

**Comment by Kully** [comment]
@chriddyp We got some nice variety going:

screen shot 2017-06-26 at 6 39 38 pm

screen shot 2017-06-26 at 6 40 04 pm

screen shot 2017-06-26 at 6 40 32 pm

**Comment by Kully** [comment]
@cldougl Can you give this one quick review today? I'm trying to get it merged today and the pip package updated so I can finish documentation by the end of the week.

**Comment by cldougl** [comment]
@Kully is it possible to change the color of the markers?
Like:
```
import plotly.plotly as py
import plotly.figure_factory as ff

import pandas as pd

mpg = pd.read_table('https://raw.githubusercontent.com/plotly/datasets/master/mpg_2017.txt')

fig = ff.create_facet_grid(
    mpg,
    x='displ',
    y='cty',
    facet_row='drv',
    facet_col='cyl',
    marker=dict(color='rgba(255, 182, 193, .9)')
)
py.iplot(fig, filename='facet_grid_mpg_two_way_facet')
```

File: plotly/figure_factory/_facet_grid.py [hunk 4]
```diff
@@ -696,6 +751,26 @@ def create_facet_grid(df, x, y, facet_row=None, facet_col=None,
...
+    import plotly.plotly as py
+    import plotly.figure_factory as ff
+
+    import pandas as pd
+
+    mtcars = pd.read_csv('https://raw.githubusercontent.com/plotly/datasets/master/mtcars.csv')
+
+    fig = ff.create_facet_grid(
+        mtcars,
+        x='wt',
```
Comment (cldougl) [inline plotly/figure_factory/_facet_grid.py]: shouldn't this example only have an x or a y?

File: plotly/figure_factory/_facet_grid.py [hunk 5]
```diff
@@ -448,35 +484,49 @@ def _facet_grid(df, x, y, facet_row, facet_col, num_of_rows,
...
-                x=group[1][x],
-                y=group[1][y],
-                mode='markers',
+            trace = dict(
                 type=trace_type,
                 marker=dict(
                     color=marker_color,
-                    **kwargs_marker
+                    line=kwargs_marker['line'],
+                    #opacity=kwargs_marker['opacity'],
```
Comment (cldougl) [inline plotly/figure_factory/_facet_grid.py]: should this be commented out?

File: plotly/figure_factory/_facet_grid.py [hunk 6]
```diff
@@ -25,6 +25,7 @@
 THRES_FOR_FLIPPED_FACET_TITLES = 10
 GRID_WIDTH = 1
 
+VALID_TRACE_TYPES = ['scatter', 'scattergl', 'histogram', 'bar', 'box']
```
Comment (cldougl) [inline plotly/figure_factory/_facet_grid.py]: are we planning on adding `histogram2d` as well (https://github.com/plotly/plotly.py/pull/776#pullrequestreview-46325756) ?

**Comment by cldougl** [comment]
screen shot 2017-06-29 at 1 58 48 pm
looks like the histogram labels will overlap due to how the ticks are labeled at the bin edges. Do you think there's something stylistically we could do to avoid this such as increasing the gap between when the histogram trace is used?

File: plotly/figure_factory/_facet_grid.py (hunk 6 above)
Comment (chriddyp) [inline plotly/figure_factory/_facet_grid.py]: and `histogram2dcontour`

**Comment by chriddyp** [comment]
looks good to me! 💃 from me once @cldougl 's comments are addressed

**Comment by Kully** [comment]
> looks like the histogram labels will overlap due to how the ticks are labeled at the bin edges. Do you think there's something stylistically we could do to avoid this such as increasing the gap between when the histogram trace is used?

Yeah

File: plotly/figure_factory/_facet_grid.py (hunk 6 above)
Comment (Kully) [inline plotly/figure_factory/_facet_grid.py]: Since `marker` is not a valid key in facet grid, I vote that these trace types get moved into another PR/Issue. It would require a large reworking of the code, so I think we should move it to another issue so we can get this merged so the documentation can go up this week.

Comment (cldougl) [inline plotly/figure_factory/_facet_grid.py]: I'm ok w/ doing those in a different pr

**Comment by cldougl** [comment]
💃 good for me too
(lmk if you want me to fix the changelog conflict @Kully )

**Comment by Kully** [comment]
> 💃 good for me too
> (lmk if you want me to fix the changelog conflict @Kully )

just did
